src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,145 @@
+
+
 package core.basesyntax;
 
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        this.head = null;
+        this.tail = null;
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        newNode.next = current;
+        newNode.prev = current.prev;
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        } else {
+            head = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
```

Comment:
The `addAll(List<T> list)` method should check if the input `list` is `null` to avoid a `NullPointerException`. Consider adding a check at the beginning of the method.